examples/src/main/java/alluxio/cli/UnderFileSystemContractTest.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.cli;
+
+import alluxio.conf.InstancedConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.examples.RelatedS3Operations;
+import alluxio.examples.S3ASpecificOperations;
+import alluxio.examples.UnderFileSystemCommonOperations;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.UnderFileSystemConfiguration;
+import alluxio.underfs.UnderFileSystemFactory;
+import alluxio.underfs.UnderFileSystemFactoryRegistry;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.util.ConfigurationUtils;
+import alluxio.util.io.PathUtils;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.UUID;
+
+/**
+ * Integration tests for Alluxio under filesystems. It describes the contract of Alluxio
+ * with the UFS through the UFS interface.
+ *
+ * This class will run all tests (with "Test" suffix in the method name)
+ * in {@link UnderFileSystemCommonOperations}. If the given ufs path is a S3 path,
+ * all tests in {@link S3ASpecificOperations} will also be run.
+ */
+public final class UnderFileSystemContractTest {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemContractTest.class);
+  private static final String S3_IDENTIFIER = "s3";
+
+  @Parameter(names = {"--path"}, required = true,
+      description = "The under filesystem path to run tests against.")
+  private String mUfsPath;
+
+  @Parameter(names = {"--help"}, help = true)
+  private boolean mHelp = false;
+
+  private InstancedConfiguration mConf
+      = new InstancedConfiguration(ConfigurationUtils.defaults());
+
+  // The factory to check if the given ufs path is valid and create ufs
+  private UnderFileSystemFactory mFactory;
+  private UnderFileSystem mUfs;
+
+  private UnderFileSystemContractTest() {}
+
+  private void run() throws Exception {
+    mFactory = UnderFileSystemFactoryRegistry.find(mUfsPath,
+        UnderFileSystemConfiguration.defaults(mConf));
+    // Check if the ufs path is valid
+    if (mFactory == null || !mFactory.supportsPath(mUfsPath)) {
+      LOG.error("{} is not a valid path", mUfsPath);
+      System.exit(1);
+    }
+
+    // Set common properties
+    mConf.set(PropertyKey.UNDERFS_LISTING_LENGTH, "50");
+    mConf.set(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT, "512B");
+    // Increase the buffer time of journal writes to speed up tests
+    mConf.set(PropertyKey.MASTER_JOURNAL_FLUSH_BATCH_TIME_MS, "1sec");
+
+    createUnderFileSystem();
+
+    runCommonOperations();
+
+    if (mUfs.getUnderFSType().equals(S3_IDENTIFIER)) {
```

Comment:
Wouldn't we want to run this for all object stores?

Reply:
The test `S3SpecificOperation` is only related to S3. Only S3 uses streaming upload. The logs inside the test are also S3 specific. The other object store can also run this test but it is not that useful and the logs may be misleading.

Reply:
Ah okay, I still would prefer a configurable way to enable this test for other object stores in case we provide these features in future.

Reply:
The other tests in the `S3ASpecificOperations` are also valid for other ufs, excepted the `createAndAbortTest` which I add a line for checking the ufs is s3. Added the `--force` that users can use to force run all the tests.